### PR TITLE
Merge ShardSearchTransportRequest and ShardSearchLocalRequest

### PR DESCRIPTION
**Pull request by jimczi**
This change merges the `ShardSearchTransportRequest` and `ShardSearchLocalRequest`
into a single `ShardSearchRequest` that can be used to create a SearchContext.

Relates #46523

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java [hunk 1]
```diff
@@ -19,81 +19,342 @@
...
-     * Returns the preference of the original {@link SearchRequest#preference()}.
-     */
-    String preference();
+    public SearchType searchType() {
+        return searchType;
+    }
 
-    /**
-     * Returns the cache key for this shard search request, based on its content
-     */
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java]: Maybe we could keep the javadocs that was there although we have a single impl?

File: server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java [hunk 2]
```diff
@@ -19,81 +19,342 @@
...
+    private final String preference;
+    private final OriginalIndices originalIndices;
 
-    ShardId shardId();
+    //these are the only two mutable fields, as they are subject to rewriting
+    private AliasFilter aliasFilter;
+    private SearchSourceBuilder source;
 
-    SearchSourceBuilder source();
+    public ShardSearchRequest(SearchRequest searchRequest,
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java]: Seems like only SliceBuilderTests uses this constructor. Can we change the caller to rely on the other constructor that accepts also the OriginalIndices argument?

**Review by javanna — APPROVED** [review]
left two minors, LGTM otherwise

File: server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java (hunk 1 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java]: good catch, thanks

File: server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java (hunk 2 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java]: +1

**Pull request merged**